test/onnx/internal/test_diagnostics.py

```diff
@@ -0,0 +1,285 @@
+# Owner(s): ["module: onnx"]
+
+import contextlib
+import dataclasses
+import io
+from typing import AbstractSet, Tuple
+
+import torch
+from torch.onnx import errors
+from torch.onnx._internal import diagnostics
+from torch.onnx._internal.diagnostics import infra
+from torch.testing._internal import common_utils
+
+
+def _assert_has_diagnostics(
+    engine: infra.DiagnosticEngine,
+    rule_level_pairs: AbstractSet[Tuple[infra.Rule, infra.Level]],
+):
+    sarif_log = engine.sarif_log()
+    unseen_pairs = {(rule.id, level.value) for rule, level in rule_level_pairs}
+    actual_results = []
+    for run in sarif_log.runs:
+        if run.results is None:
+            continue
+        for result in run.results:
+            id_level_pair = (result.rule_id, result.level)
+            unseen_pairs.discard(id_level_pair)
+            actual_results.append(id_level_pair)
+
+    if unseen_pairs:
+        raise AssertionError(
+            f"Expected diagnostic results of rule id and level pair {unseen_pairs} not found. "
+            f"Actual diagnostic results: {actual_results}"
+        )
+
+
+@contextlib.contextmanager
+def assert_all_diagnostics(
+    test_suite: common_utils.TestCase,
+    engine: infra.DiagnosticEngine,
+    rule_level_pairs: AbstractSet[Tuple[infra.Rule, infra.Level]],
+):
+    """
+    Context manager to assert that all diagnostics are emitted.
+
+    Usage:
+        with assert_all_diagnostics(
+            self,
+            diagnostics.engine,
+            {(rule, infra.Level.Error)},
+        ):
+            torch.onnx.export(...)
+
+    Args:
+        test_suite: The test suite instance.
+        engine: The diagnostic engine.
+        rule_level_pairs: A set of rule and level pairs to assert.
+
+    Returns:
+        A context manager.
+
+    Raises:
+        AssertionError: If not all diagnostics are emitted.
+    """
+
+    try:
+        yield
+    except errors.OnnxExporterError:
+        test_suite.assertIn(infra.Level.ERROR, {level for _, level in rule_level_pairs})
+    finally:
+        _assert_has_diagnostics(engine, rule_level_pairs)
+
+
+def assert_diagnostic(
+    test_suite: common_utils.TestCase,
```

Comment:
```suggestion
    test_suite: unittest.TestCase,
```

nit, optional: Can it work with arbitrary unittest test cases?

Reply:
I think so, updated.